nixos/modules/services/networking/unbound.nix

```diff
@@ -89,86 +45,136 @@ in
-        type = types.listOf types.str;
-        description = ''
-          What addresses/interfaces are allowed to remote-control the unbound server.
-          Requires <literal>enableRemoteAccess</literal> set to true.
+      settings = mkOption {
+        default = {};
+        type = with types; submodule {
+
+          freeformType = let
+            validSettingsPrimitiveTypes = oneOf [ int str bool float ];
+            validSettingsTypes = oneOf [ validSettingsPrimitiveTypes (listOf validSettingsPrimitiveTypes) ];
+            settingsType = (attrsOf validSettingsTypes);
+          in attrsOf (oneOf [ settingsType (listOf settingsType) ])
+              // { description = ''
+                unbound.conf configuration type. The format consist of an attribute
+                set of settings. Each settings can be either one value, a list of
+                values or an attribute set. The allowed values are integers,
+                strings, booleans or floats.
+              '';
+            };
+
+          options = {
+            remote-control.control-enable = mkOption {
+              type = bool;
+              default = false;
+              description = "";
+            };
+          };
+       };
+        example = literalExample ''
+          {
+            server = {
+              interface = [ "127.0.0.1" ];
+            };
+            forward-zone = [
+              {
+                name = ".";
+                forward-addr = "1.1.1.1@853#cloudflare-dns.com";
+              }
+              {
+                name = "example.org.";
+                forward-addr = [
+                  "1.1.1.1@853#cloudflare-dns.com"
+                  "1.0.0.1@853#cloudflare-dns.com"
+                ];
+              }
+            ];
+            remote-control.control-enable = true;
+          };
         '';
-      };
-
-      remoteAccessPort = mkOption {
-        default = 8953;
-        type = types.port;
         description = ''
-          What port the unbound server will listen on for remote-control.
-          Requires <literal>enableRemoteAccess</literal> set to true.
+          Declarative Unbound configuration
+          See the <citerefentry><refentrytitle>unbound.conf</refentrytitle>
+          <manvolnum>5</manvolnum></citerefentry> manpage for a list of
+          available options.
         '';
       };
-
     };
   };
 
   ###### implementation
 
   config = mkIf cfg.enable {
 
-    environment.systemPackages = [ cfg.package (hiPrio unboundWrapped) ];
+    services.unbound.settings = {
+      server = {
+        directory = mkDefault cfg.stateDir;
+        username = cfg.user;
+        chroot = cfg.stateDir;
+        pidfile = ''""'';
+        interface = mkDefault ([ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1"));
+        access-control = mkDefault [ "127.0.0.0/24 allow" ];
+        auto-trust-anchor-file = mkDefault (if cfg.enableRootTrustAnchor then rootTrustAnchorFile else null);
```

Comment:
When `enableRootTrustAnchor` is `false`:

    error: The option value `services.unbound.settings.server.auto-trust-anchor-file' in `$nixpkgs/nixos/modules/services/networking/unbound.nix' is not of type `signed integer or string or boolean or floating point number or list of signed integer or string or boolean or floating point numbers'.

Reply:
@wahjava could you try again with my latest modifications?

Reply:
Works as expected now, thanks!